lambda/translattribute.ml

```diff
@@ -333,6 +378,49 @@ let add_loop_attribute expr loc attributes =
     end
   | _ -> expr
 
+let add_regalloc_attribute expr loc attributes =
+  match expr with
+  | Lfunction({ attr = { stub = false } as attr } as funct) ->
+    begin match get_regalloc_attribute attributes with
+    | None -> expr
+    | (Some _) as regalloc ->
+      begin match attr.regalloc with
+      | None -> ()
+      | Some _ ->
+          Location.prerr_warning loc
+            (Warnings.Duplicated_attribute "regalloc")
+      end;
+      let attr = { attr with regalloc } in
+      lfunction_with_attr ~attr funct
+    end
+  | _ -> expr
+
+let add_regalloc_param_attribute expr _loc attributes =
+  match expr with
+  | Lfunction({ attr } as funct) ->
+    let params = get_regalloc_param_attributes attributes in
+    begin match params with
+    | [] -> expr
+    | _ ->
+      let attr = { attr with regalloc_param = attr.regalloc_param @ params } in
+      lfunction_with_attr ~attr funct
+    end
+  | _ -> expr
+
+let add_cold_attribute expr loc attributes =
+  match expr with
+  | Lfunction({ attr = { stub = false } as attr } as funct) ->
+    if get_cold_attribute attributes then begin
+      if attr.cold then Location.prerr_warning loc
+            (Warnings.Duplicated_attribute "cold");
+      (* ppx_cold rewrites `[@cold]` to `[@inline never][@specialise never][@local never]`
+         so we do the equivalent here. *)
+      let attr = { attr with cold = true; inline = Never_inline; specialise = Never_specialise; local = Never_local; } in
```

Comment:
should there be another duplicate warning here if `inline` and friends are not `Default`?